Sources/SwiftSyntaxMacros/MacroProtocols/BodyMacro.swift

```diff
@@ -0,0 +1,26 @@
+//
+// Copyright (c) 2014 - 2023 Apple Inc. and the Swift project authors
+// Licensed under Apache License v2.0 with Runtime Library Exception
+//
+// See https://swift.org/LICENSE.txt for license information
+// See https://swift.org/CONTRIBUTORS.txt for the list of Swift project authors
+//
+//===----------------------------------------------------------------------===//
+
+import SwiftSyntax
+import SwiftSyntaxBuilder  // FIXME: Why is HasTrailingOptionalCodeBlock there?
+
+/// Describes a macro that can create the body for a function that does not
+/// have one.
+public protocol BodyMacro: AttachedMacro {
+  /// Expand a macro described by the given custom attribute and
+  /// attached to the given declaration and evaluated within a
+  /// particular expansion context.
+  ///
+  /// The macro expansion can introduce a body for the given function.
+  static func expansion(
+    of node: AttributeSyntax,
+    providingBodyFor declaration: some DeclSyntaxProtocol & HasTrailingOptionalCodeBlock,
```

Comment:
Why do we need `HasTrailingOptionalCodeBlock` here? If the body macro adds a body, it shouldn’t really care if `declaration` already contains a body.

Reply:
If the declaration already contains a body, the macro is likely going to want to incorporate that body into the new body it generates. For example, wrapping the body as written into a closure that's passed to some other function that does some additional work around the call to the closure.